### DISCUSSION TITLE
Emit spans which show exponential backoff delay and jitter

### DISCUSSION THREAD

**Discussion by lopopolo-openai** [hynek/stamina#84] [Ideas]
I have a function that internally uses `stamina.retry_context` to retry a cross service network call. We use tracing on these requests to allow perf debugging. I've attached below the spans that get emitted by this code. I am fairly certain the gaps between the spans here at least in part are windows where stamina is doing exponential backoff and jitter between the retries, but I am not sure if there is other stuff going on in there too.

<img width="724" alt="Screenshot 2024-12-05 at 12 40 29 PM" src="https://github.com/user-attachments/assets/e151e8be-6e0a-43aa-9f8b-165458952be6">

Would it be possible to take a callback or something that can be used to get stamina to emit spans for the relevant delays in the retry cycle? It would be excellent if these spans included info about the amount of backoff and jitter, and what the next attempt number will be.

**Comment by hynek** [comment]
I'm not 100% positive on what you mean by that, but what is missing from https://stamina.hynek.me/en/stable/api.html#stamina.instrumentation.RetryDetails to implement a retry hook?

**Reply by hynek** [reply]
@lopopolo-openai btw it would be good (for you :)) if we could resolve this fast-ish, because I'm in the last leg of editing a video about stamina and want to release a new version at the same time. :)

**Reply by lopopolo-openai** [reply]
I'll take a peek on Monday. Usually our tracing code expects spans to be tracked using a context manager.

**Reply by hynek** [reply]
hmmm, I might be convinced to allow for retry hooks to be context managers but that probably is nothing I can very short term 🤔

**Comment by hynek** [comment]
fixed in #95
